src/validation.cpp

```diff
@@ -465,6 +463,114 @@ static bool CheckInputsFromMempoolAndCache(const CTransaction& tx, TxValidationS
     return CheckInputScripts(tx, state, view, flags, /* cacheSigStore = */ true, /* cacheFullSciptStore = */ true, txdata);
 }
 
+/** An empty coin used as a placeholder for a spent coin.*/
+static const Coin coin_spent;
+/**
+ * A CoinsView that adds a memory cache to another CoinsView and serves as temporary scratch space.
+ * Used by MemPoolAccept class to validate transactions and packages before submitting to mempool.
+ * A backend can be set to provide read access to chainstate and/or mempool coins, but writing to
+ * the backend is disabled. Avoid using a CCoinsViewTemporary in consensus-critical paths such
+ * as writing to the script cache. See CheckInputsFromMempoolAndCache as an example. When not being
+ * used to validate a package (m_temp_added and m_temp_spent are empty), a CCoinsViewTemporary
+ * behaves exactly like a CCoinsViewCache.
+ */
+class CCoinsViewTemporary : public CCoinsViewCache
+{
+protected:
+    /**
+    * Coins made available by transactions being validated. Tracking these allows for package
+    * validation, since we can access transaction outputs without submitting them to mempool.
+    */
+    std::map<COutPoint, Coin> m_temp_added;
+
+    /**
+    * Coins spent by transactions being validated. When there are multiple, we need to track these
+    * in order to distinguish between missing/spent coins and conflicts within a package.
+    */
+    std::set<COutPoint> m_temp_spent;
```

Comment:
Thanks for the suggestion! I considered this when I first saw the comment, but since then, I've updated `AcceptMultipleTransactions()` to just remove `m_temp_spent` after all the `PreChecks()` so that the script checks can use coins from `m_temp_added`. That makes this suggestion no longer apply so I'm going to mark it as resolved.